### PR TITLE
AP_OSD: add power and energy items

### PR DESCRIPTION
**Pull request by shellixyz**

### CODE REVIEW

**Comment by shellixyz** [comment]
Updated

**Comment by kiryam** [comment]
👍

File: libraries/AP_OSD/AP_OSD_Screen.cpp [hunk 1]
```diff
@@ -704,7 +704,39 @@ const AP_Param::GroupInfo AP_OSD_Screen::var_info[] = {
     // @Description: Vertical position on screen
     // @Range: 0 15
     AP_SUBGROUPINFO(clk, "CLK", 43, AP_OSD_Screen, AP_OSD_Setting),
-    
+
+    // @Param: POWER_EN
+    // @DisplayName: POWER_EN
+    // @Description: Displays total drawn current
```
Comment (Hwurzburg) [inline libraries/AP_OSD/AP_OSD_Screen.cpp:710]: should be "Displays total power draw from all batteries", not current

File: libraries/AP_OSD/AP_OSD_Screen.cpp [hunk 2]
```diff
@@ -704,7 +704,39 @@ const AP_Param::GroupInfo AP_OSD_Screen::var_info[] = {
...
+
+    // @Param: POWER_Y
+    // @DisplayName: POWER_Y
+    // @Description: Vertical position on screen
+    // @Range: 0 15
+    AP_SUBGROUPINFO(power, "POWER", 44, AP_OSD_Screen, AP_OSD_Setting),
+
+    // @Param: ENERGY_EN
+    // @DisplayName: ENERGY_EN
+    // @Description: Displays main consumed energy
```
Comment (Hwurzburg) [inline libraries/AP_OSD/AP_OSD_Screen.cpp:726]: perhaps "Displays total energy consumed from primary battery" ?

**Comment by shellixyz** [comment]
Changed param descriptions as suggested

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp [hunk 3]
```diff
@@ -428,6 +428,29 @@ void AP_BattMonitor::check_failsafes(void)
...
+    for (uint8_t instance = 0; instance < _num_instances; instance++) {
+        float instance_power;
+        if (power_watts(instance_power, instance)) {
+            total += instance_power;
+        }
+    }
+    return total;
+}
+
+bool AP_BattMonitor::power_watts(float &power, uint8_t instance) const
```
Comment (peterbarker) [inline libraries/AP_BattMonitor/AP_BattMonitor.cpp:444]: Instance number first by convention.

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp [hunk 4]
```diff
@@ -428,6 +428,29 @@ void AP_BattMonitor::check_failsafes(void)
...
+{
+    if ((instance < _num_instances) && (drivers[instance] != nullptr) && drivers[instance]->has_current()) {
+        power = state[instance].current_amps * state[instance].voltage;
+        return true;
+    } else {
```
Comment (peterbarker) [inline libraries/AP_BattMonitor/AP_BattMonitor.cpp:449]: Don't need this `else`

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp (hunk 3 above)
Comment (shellixyz) [reply libraries/AP_BattMonitor/AP_BattMonitor.cpp:444]: By convention ? I just copied the signatures of `AP_BattMonitor::consumed_mah`, `AP_BattMonitor::current_amps`, ...

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp (hunk 4 above)
Comment (shellixyz) [reply libraries/AP_BattMonitor/AP_BattMonitor.cpp]: Fixed

File: libraries/AP_BattMonitor/AP_BattMonitor.h [hunk 5]
```diff
@@ -144,6 +144,10 @@ class AP_BattMonitor
     /// current_amps - returns the instantaneous current draw in amperes
     bool current_amps(float &current, const uint8_t instance = AP_BATT_PRIMARY_INSTANCE) const WARN_IF_UNUSED;
 
+    /// power watt
+    float power_watts() const;
+    bool power_watts(float &power, const uint8_t instance = AP_BATT_PRIMARY_INSTANCE) const;
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor.h:149]: No default to the primary instance, all the primary instance stuff in battery monitor is on the chopping block, so lets not extend it in the meantime.

**Review by WickedShell — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
I think it makes sense to have a OSD_BAT_IDX parameter for which battery index to use for all the battery based OSD elements

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp [hunk 6]
```diff
@@ -428,6 +428,29 @@ void AP_BattMonitor::check_failsafes(void)
     }
 }
 
+// returns the total power draw for all batteries
+float AP_BattMonitor::power_watts() const
```
Comment (tridge) [inline libraries/AP_BattMonitor/AP_BattMonitor.cpp:432]: summing causes issues with some battery types (fuel systems, already summed batteries etc)
better to add a OSD_BAT_IDX param

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by skorokithakis** [comment]
I'd like to +1 this PR, measuring power/energy is much better than mAh and OSD items for them would be very useful.

**Pull request closed without merging**